compiler/rustc_hir_analysis/src/check/compare_eii.rs

```diff
@@ -0,0 +1,400 @@
+//! This module is very similar to `compare_impl_item`.
+//! Most logic is taken from there,
+//! since in a very similar way we're comparing some declaration of a signature to an implementation.
+//! The major difference is that we don't bother with self types, since for EIIs we're comparing freestanding item.
+
+use std::borrow::Cow;
+use std::iter;
+
+use rustc_data_structures::fx::FxIndexSet;
+use rustc_errors::{Applicability, E0806, struct_span_code_err};
+use rustc_hir::def_id::{DefId, LocalDefId};
+use rustc_hir::{self as hir, FnSig, HirId, ItemKind};
+use rustc_infer::infer::{self, InferCtxt, TyCtxtInferExt};
+use rustc_infer::traits::{ObligationCause, ObligationCauseCode};
+use rustc_middle::ty::error::{ExpectedFound, TypeError};
+use rustc_middle::ty::{self, TyCtxt, TypeVisitableExt, TypingMode};
+use rustc_span::{ErrorGuaranteed, Ident, Span, Symbol};
+use rustc_trait_selection::error_reporting::InferCtxtErrorExt;
+use rustc_trait_selection::regions::InferCtxtRegionExt;
+use rustc_trait_selection::traits::{self, ObligationCtxt};
+use tracing::{debug, instrument};
+
+use super::potentially_plural_count;
+use crate::check::compare_impl_item::{CheckRegionBoundsOnItemOutput, check_region_bounds_on_item};
+use crate::errors::{EiiWithGenerics, LifetimesOrBoundsMismatchOnEii};
+
+/// checks whether the signature of some `external_impl`, matches
+/// the signature of `declaration`, which it is supposed to be compatible
+/// with in order to implement the item.
+pub(crate) fn compare_eii_function_types<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    external_impl: LocalDefId,
+    declaration: DefId,
+    eii_name: Symbol,
+    eii_attr_span: Span,
+) -> Result<(), ErrorGuaranteed> {
+    check_is_structurally_compatible(tcx, external_impl, declaration, eii_name, eii_attr_span)?;
+
+    let external_impl_span = tcx.def_span(external_impl);
+    let cause = ObligationCause::new(
+        external_impl_span,
+        external_impl,
+        ObligationCauseCode::CompareEii { external_impl, declaration },
+    );
+
+    // no trait bounds
+    let param_env = ty::ParamEnv::empty();
+
+    let infcx = &tcx.infer_ctxt().build(TypingMode::non_body_analysis());
+    let ocx = ObligationCtxt::new_with_diagnostics(infcx);
+
+    // We now need to check that the signature of the impl method is
+    // compatible with that of the trait method. We do this by
+    // checking that `impl_fty <: trait_fty`.
+    //
+    // FIXME: We manually instantiate the trait method here as we need
+    // to manually compute its implied bounds. Otherwise this could just
+    // be ocx.sub(impl_sig, trait_sig).
+
+    let wf_tys = FxIndexSet::default();
```

Comment:
we're not actually computing implied bounds here xd

Either change the `FIXME` to state that we're not computing implied bounds, or start computing the implied bounds. Also make sure all the comments here talk about "declaration" and "external_impl" instead of trait vs impl